### DISCUSSION TITLE
Fastapi plugin: lifespan doesn't persists yield output to state

### DISCUSSION THREAD

**Discussion by OriBerlo** [ag2ai/faststream#1653] [General]
Hi,
I'm using faststream with the Fastapi plugin. I'm using your suggestion about setting a lifespan to the rabbit router, my issue starts when I try to fully utilize the existing fastapi capabilities to pass state in the requests object.

here are some code samples:

```python
from faststream.rabbit.fastapi import RabbitRouter

broker_endpoint = communications_config.rabbitmq.broker_endpoint
rabbit_router = RabbitRouter(broker_endpoint)
```

```python
@asynccontextmanager
async def qs_lifespan(app: FastAPI) -> AsyncGenerator[Dict[str, Any], None]:
    async with (rabbit_router.lifespan_context(app)):
        config =load_config() # My utility which load configuration
        yield {"config": config}

if __name__ == "__main__":
    app = FastAPI(lifespan=qs_lifespan)
    app.include_router(rabbit_router)
```

took the liberty to debug and compare the lifespan yield process between a regular HTTP fastapi and my app when incorporating a faststream router and seems that In the starlette/routing.py (return function of the lifespan) both acquire the state the same way.

another small question, the RabbitRouter starts serving rabbit mq connection the moment it initialized, is there any option of delaying its start after lifespan is finished similar to how Fastapi routers work?

I'd be glad if someone can assist me with my issue
Thanks!

**Comment by kumaranvpl** [comment]
Hello @OriBerlo, 

I have failed to understand your problem because of the language. Could you please try to elaborate more?

**Reply by OriBerlo** [reply]
Hi @kumaranvpl, 
I have an issue with utilizing fastapi features when using faststream with the fastapi plugin.

1. Fastapi has the ability of yielding values from the lifespan, this values are populated in the starlette.request under state property.
when yielding a dictionary for faststream router the yielded value is not populated under request.state.{value} like I would expect from fastapi app.

2. Faststream is initialized and listens for connection the moment it's declared (RabbitRouter("URL")). is there a way to make a rabbit router start listening only when the lifespan is finished? (in fastapi the routers begin listening only after the lifespan is finished).


hope I simplified the question a bit, but reach out if anything remain unclear.
